### PR TITLE
log exceptions before killing the server

### PR DESCRIPTION
**Pull request by pepeiborra**
I believe that this will help understand the "resource vanished - broken pipe" test failures

### CODE REVIEW

File: ghcide/src/Development/IDE/LSP/LanguageServer.hs [hunk 1]
```diff
@@ -141,7 +141,10 @@ runLanguageServer options inH outH getHieDbLoc defaultConfig onConfigurationChan
             logInfo (ideLogger ide) $ T.pack $ "Registering ide configuration: " <> show initConfig
             registerIdeConfiguration (shakeExtras ide) initConfig
 
-            _ <- flip forkFinally (const exitClientMsg) $ runWithDb dbLoc $ \hiedb hieChan -> do
+            let handleHieDbException e = do
+                    logError (ideLogger ide) (T.pack $ "Unexpected exception in hiedb thread: " <> show e)
+                    exitClientMsg
+            _ <- flip forkFinally handleHieDbException $ runWithDb dbLoc $ \hiedb hieChan -> do
```
Comment (wz1000) [inline ghcide/src/Development/IDE/LSP/LanguageServer.hs:147]: This isn't the hiedb thread, but the server thread. The hiedb thread is forked again in `runWithDb` and shouldn't bring down the main server thread.

Comment (pepeiborra) [reply ghcide/src/Development/IDE/LSP/LanguageServer.hs]: Thanks, I didn't look closely enough. I have fixed the log message to reflect that this is the server thread

**Review by jneira — APPROVED** [review]
thanks, lets see what can it tell us

File: ghcide/src/Development/IDE/LSP/LanguageServer.hs [hunk 2]
```diff
@@ -141,7 +141,10 @@ runLanguageServer options inH outH getHieDbLoc defaultConfig onConfigurationChan
             logInfo (ideLogger ide) $ T.pack $ "Registering ide configuration: " <> show initConfig
             registerIdeConfiguration (shakeExtras ide) initConfig
 
-            _ <- flip forkFinally (const exitClientMsg) $ runWithDb dbLoc $ \hiedb hieChan -> do
+            let handleServerException e = do
+                    logError (ideLogger ide) (T.pack $ "Unexpected exception in server thread: " <> show e)
+                    exitClientMsg
+            _ <- flip forkFinally handleServerException $ runWithDb dbLoc $ \hiedb hieChan -> do
```
Comment (berberman) [inline ghcide/src/Development/IDE/LSP/LanguageServer.hs:147]: `handleServerException` always runs after the action. We should check `e` is `Left SomeException` or `Right ()`.

**Pull request merged**